### ISSUE TITLE
Next-page stops working after panning in fit page width and fit content modes.

### ISSUE THREAD

**Issue by tigran123** [koreader/kindlepdfviewer#236]
The file test/tall.pdf is only a single page which is probably why this bug was not noticed before, so you need to do something like "pdftk tall.pdf tall.pdf tall.pdf tall.pdf cat output tall4p.pdf" and test on tall4p.pdf to reproduce it.
1. Open tall4p.pdf on page 1
2. Press Shift-S to switch to "fit content width" (or S for "fit page width" --- results are the same)
3. Use joypad to pan the view down a bit.
4. Press "Next page" button: what you will see on the screen will be exactly the same as before pressing it (because all pages in tall4.pdf are identical), but pressing "G" you find that the page is now number 2, i.e. the page was turned instead of moving screen full down (as before panning).

I expect "Next page" to behave after panning in the same way as before panning, i.e. move down in large chunks (almost screen full). This is because panning is useful in situations where some interesting element (picture or equation) got split between screens and you pan down a bit to shift it to the visible area. After viewing the element you then desire to go on reading as before, so you press "Next page", but, unexpectedly, it starts jumping physical pages...

Also, if I patiently pan the view down to the very bottom of the page and then press "Next page" I expect to see the _top_ of the next page --- instead I see the _bottom_ thereof. And pressing "Next page" again shows me the bottom of page 3 and another "Next page" the bottom of page 4. (pressing "G" I ascertain the current page number).

**Comment by tigran123** [comment]
Just wanted to add that the bug is reproducible on other tall files too, not just the above artificially constructed tall4p.pdf. I.e. I originally noticed it while reading normal books.

**Comment by dpavlin** [comment]
I had similar dilemma when implementing two column mode, but I decided not to change prev/next keys meaning and leave them for page flipping always to have consistent behavior of page keys in all modes.

However, crengine is behaving in the way you are describing (next page gives you next part of document) and I can see benefits of both approaches.

To make thing even more interesting, if we had continuous page scroll described in #190 page panning alone would provide desired effect.

Simplest solution which comes to mind is to change how page keys work in zoom modes which are not fit to screen: next page should reset to top of page and previous to bottom similar to two column mode.

Does any of this make sense, or did I just wrote a huge brain dump?

**Comment by tigran123** [comment]
The behaviour of prev/next keys is not always "flip to the prev/next page". If you pan the view down a bit (using 5-way controller down) then the behaviour of "next" switches to "flip to the next page", but _before_ panning the behaviour of "next" is "pan down one screen full".

The "next" page should reset to top of next page only if there is nothing more to be displayed on the current page. But if there is still something then it should pan down one screen-full, just like it did prior to using 5-way controller.

Otherwise, please tell me: how do you expect someone to continue reading after using the 5-way controller in the middle of the page? Currently, the _only_ way that I could find (to workaround the problem) is to reset the view ('A' command) and then switch to "fit to content" (Shift-S) mode again and then keep pressing "next" page enough times to get to where I want to be. Surely this is not normal?

**Comment by dpavlin** [comment]
I totally agree that it doesn't make much sense.

Part of the problem is that reader is modal -- it's in `ZOOM_FIT_TO_CONTENT_WIDTH` or `ZOOM_FIT_TO_CONTENT_WIDTH_PAN` mode and user doesn't have any indication in which mode it is.

So, when reader switches from `ZOOM_FIT_TO_CONTENT_WIDTH_PAN` to `ZOOM_FIT_TO_CONTENT_WIDTH` on fiveway press, it's enough to press just `Shift+S` (without need for `A`) to go into `ZOOM_FIT_TO_CONTENT_WIDTH_PAN`.

We really need some kind of indication where page borders are relative to current view of page (this will help a lot when you are changing pages, panning and you don't know where in document you are) and indicator in which zoom mode are you currently. I have some ideas about that, but no code to show (yet ;-)

You are welcomed to propose better or simplifier solution to this problem.

**Comment by tigran123** [comment]
Thank you, I get it now.

Ok, in that case, yes, maybe we could do this:
1. When we press Shift-S after having panned with a fiveway the view should not be reset to the top but stay where it is.
2. As you say, some kind of indicator (but where to put it? space is so precious...) should tell us which mode we are in.

I would be against using any space that could otherwise be used for the useful data, i.e. for the page itself. Hmmm, maybe some grey overlay at the side (like on Nook Touch jailbroken with proper Android running), with the ability to turn this indicator off? Not sure... But I understand why the behaviour is the way it is --- thank you, @dpavlin for taking the time to explain it.

**Comment by dpavlin** [comment]
AD 1: that would make sense. After playing around with multi-page tall file, I think we should also do positioning on top/bottom of page when you press prev/next page as you suggested before.

AD 2: FWIW, I also dislike indicators which take space, but showing it for 2 seconds after page refresh and/or when you press menu might be good enough solution. I'm thinking about drawing it with xor, so changes can easily be reverted.

**Comment by ghost** [comment]
Just info: I use a bit another way to see which zoom-mode is active. Since I always forget which hotkey is responcible for which mode, I just replaced 8 hotkeys (i.e. A, ^A, S, ^S, D, ^D, F and ^F) by the menu where autoselected item points to self.globalzoom_mode. If I do not want to switch reader to another mode, I just press Back. 

self.commands:add(KEY_M, MOD_ALT, "M",
        "select zoom mode",
        function(unireader)
            local mode_list = { --                  VALUE                           OLD HOTKEY
                "Enter zoom manually",              -- ZOOM_BY_VALUE = 0,
                "Fit zoom to page",                 -- ZOOM_FIT_TO_PAGE = -1,               A
                "Fit zoom to page width",               -- ZOOM_FIT_TO_PAGE_WIDTH = -2,         S
                "Fit zoom to page height",              -- ZOOM_FIT_TO_PAGE_HEIGHT = -3,            D
                "Fit zoom to content",              -- ZOOM_FIT_TO_CONTENT = -4,            ^A
                "Fit zoom to content width",            -- ZOOM_FIT_TO_CONTENT_WIDTH = -5,          ^S
                "Fit zoom to content height",           -- ZOOM_FIT_TO_CONTENT_HEIGHT = -6,     ^D
                "Fit zoom to content width with panoraming",    -- ZOOM_FIT_TO_CONTENT_WIDTH_PAN = -7,
                "Fit zoom to content height with panoraming",   -- ZOOM_FIT_TO_CONTENT_HEIGHT_PAN = -8,
                "Fit zoom to content half-width with margin",   -- ZOOM_FIT_TO_CONTENT_HALF_WIDTH_MARGIN = -9,  F
                "Fit zoom to content half-width",           -- ZOOM_FIT_TO_CONTENT_HALF_WIDTH = -10,        ^F
                }

```
        local zoom_menu = SelectMenu:new{
            menu_title = "Select mode to zoom pages",
            item_array = mode_list,
            current_entry = -self.globalzoom_mode,
            }
        local re, mode = zoom_menu:choose(0, fb.bb:getHeight())
        if re then
            --InfoMessage:show(string.sub(mode,6,-1).." = "..(1-re), 2000, 1)
            unireader:setglobalzoom_mode(1-re)
        else
            self:redrawCurrentPage()
        end
    end
)
```

**Comment by dpavlin** [comment]
During testing of this change I found bug when we are on bottom of the page and press fiveway up, part of page becomes centered on screen instead of flushed to top. This sill has to be fixed.

**Comment by tigran123** [comment]
Pull request with the fix #329 was merged in, so closing the issue.
